File: backend/tools/rag_tools.py

from langchain_core.tools import tool

# Global retriever instance (set by RAG agent during initialization)
_retriever = None


def set_retriever(retriever):
    """Set the retriever instance for the tool to use.

    Args:
        retriever: LangChain retriever instance
    """
    global _retriever
    _retriever = retriever


@tool
def retriever_tool(query: str) -> str:
    """Search and return information from the strength and conditioning gym documents.

    Args:
        query: The search query

    Returns:
        Retrieved information from gym documents
    """
    if _retriever is None:
        raise RuntimeError("Retriever not initialized. Call set_retriever() first.")

    docs = _retriever.invoke(query)

    if not docs:
        return "No relevant information found."

    results = []
    for i, doc in enumerate(docs):
        results.append(f"Document {i+1}:\n{doc.page_content}\n")

    return "\n\n".join(results)

File: backend/src/tools/ragTools.ts
/**
 * RAG tools for the RAGAgent.
 *
 * The retriever must be initialized via `setRetriever` before use.
 */

import { DynamicStructuredTool } from "@langchain/core/tools";
import { z } from "zod";
import type { VectorStoreRetriever } from "@langchain/core/vectorstores";

let retriever: VectorStoreRetriever | null = null;

/**
 * Injects the vector store retriever for document search.
 *
 * Must be called during initialization before the RAGAgent processes requests.
 */
export function setRetriever(r: VectorStoreRetriever): void {
  retriever = r;
}

export const retrieverTool = new DynamicStructuredTool({
  name: "retriever_tool",
  description:
    "Search and return information from the strength and conditioning gym documents",
  schema: z.object({
    query: z.string().describe("The search query"),
  }),
  func: async ({ query }) => {
    if (!retriever) {
      throw new Error("Retriever not initialized. Call setRetriever() first.");
    }

    const docs = await retriever.invoke(query);

    if (!docs.length) {
      return "No relevant information found.";
    }

    return docs.map((doc, i) => `Document ${i + 1}:\n${doc.pageContent}\n`).join("\n\n");
  },
});

/** All RAG tools available to the RAGAgent. */
export const ragTools = [retrieverTool];
